pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/FileUploadPathProvider.java

```diff
@@ -30,79 +30,79 @@
 
 // TODO This is a misc class used by many jersey apis. Need to rename it correctly
 public class FileUploadPathProvider {
+  private static final Logger LOGGER = LoggerFactory.getLogger(FileUploadPathProvider.class);
+
   public final static String STATE = "state";
   public final static String TABLE_NAME = "tableName";
-  private final static String DEFAULT_SCHEME = "file";
+
   private final static String FILE_UPLOAD_TEMP_PATH = "/fileUploadTemp";
   private final static String UNTARRED_PATH = "/untarred";
   private final static String SCHEMAS_TEMP = "/schemasTemp";
-  private static final Logger LOGGER = LoggerFactory.getLogger(FileUploadPathProvider.class);
-
-  private final ControllerConf _controllerConf;
 
-  private final URI _fileUploadTmpDirURI;
   private final URI _baseDataDirURI;
-  private final URI _tmpUntarredPathURI;
   private final URI _schemasTmpDirURI;
   private final URI _localTempDirURI;
+  private final URI _fileUploadTmpDirURI;
+  private final URI _tmpUntarredPathURI;
   private final String _vip;
 
   public FileUploadPathProvider(ControllerConf controllerConf) throws InvalidControllerConfigException {
-    _controllerConf = controllerConf;
-
-    // Pick the correct scheme for PinotFS. For pluggable storage, we will expect that the directory is configured
-    // with a storage prefix. For backwards compatibility, we will assume local/mounted NFS storage if nothing is configured.
-    String scheme = DEFAULT_SCHEME;
-    String dataDir = _controllerConf.getDataDir();
+    String dataDir = controllerConf.getDataDir();
     try {
-      URI uri = new URI(_controllerConf.getDataDir());
-      String dataDirScheme = uri.getScheme();
-      if (dataDirScheme ==  null) {
-        // Assume local fs
-        dataDir = new URI(CommonConstants.Segment.LOCAL_SEGMENT_SCHEME, _controllerConf.getDataDir(), null).toString();
+      // URIs that are allowed to be remote
+      _baseDataDirURI = getUriFromPath(dataDir);
+      LOGGER.info("Data directory: {}", _baseDataDirURI);
+      _schemasTmpDirURI = new URI(_baseDataDirURI + SCHEMAS_TEMP);
+      LOGGER.info("Schema temporary directory: {}", _schemasTmpDirURI);
+      String scheme = _baseDataDirURI.getScheme();
+      PinotFS pinotFS = PinotFSFactory.create(scheme);
+      mkdirIfNotExists(pinotFS, _baseDataDirURI);
+      mkdirIfNotExists(pinotFS, _schemasTmpDirURI);
+
+      // URIs that are always local
+      String localTempDir = controllerConf.getLocalTempDir();
+      if (localTempDir == null) {
+        LOGGER.info("Local temporary directory is not configured, use data directory as the local temporary directory");
+        _localTempDirURI = _baseDataDirURI;
       } else {
-        scheme = dataDirScheme;
+        _localTempDirURI = getUriFromPath(localTempDir);
```

Comment:
What input are you planning on allowing?
I see that normal directories are allowed but curious about different file:// prefixes? Do all amounts of slashes work?

Reply:
None of them works except for file://. There is a check just after this